### PR TITLE
Refactoring local factory

### PR DESCRIPTION
**Pull request by TomKlotzPro**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | no
| New feature?      | yes   
| BC breaks?        | no     
| Related Issue     | [#150](https://github.com/algolia/scout-extended/issues/150)
| Need Doc update   | no

## Describe your change

Refactoring LocalFactory by adding SettingContract(interface) and Setting class files. 

This is related to this issue :[#150](https://github.com/algolia/scout-extended/issues/150)

### CODE REVIEW

**Comment by TomKlotzPro** [comment]
you can review @nunomaduro

File: src/Contracts/SettingContract.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ *  For the full copyright and license information, please view the LICENSE
+ *  file that was distributed with this source code.
+ */
+
+namespace Algolia\ScoutExtended\Contracts;
+
+interface SettingContract
+{
+    public static function exist(string $key, $value, array $array): array;
```
Comment (nunomaduro) [inline src/Contracts/SettingContract.php]: Please refactor the code in the way, that `only` if exist you set the value.

Comment (nunomaduro) [inline src/Contracts/SettingContract.php]: Add docs on the interface.

Comment (nunomaduro) [reply src/Contracts/SettingContract.php]: So you should have: `exist` and `getValue`.

File: src/Contracts/SettingContract.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * (c) Algolia Team <contact@algolia.com>
+ *
+ *  For the full copyright and license information, please view the LICENSE
+ *  file that was distributed with this source code.
+ */
+
+namespace Algolia\ScoutExtended\Contracts;
+
+interface SettingContract
```
Comment (nunomaduro) [inline src/Contracts/SettingContract.php]: Add `@internal` phpdocs.

File: src/Contracts/SettingContract.php (hunk 1 above)
Comment (nunomaduro) [inline src/Contracts/SettingContract.php]: getValue();

File: src/Settings/LocalFactory.php [hunk 3]
```diff
@@ -128,117 +70,26 @@ public function __construct(RemoteSettingsRepository $remoteRepository)
     public function create(SearchIndex $index, string $model): Settings
     {
         $attributes = $this->getAttributes($model);
-        $searchableAttributes = [];
-        $attributesForFaceting = [];
-        $customRanking = [];
-        $disableTypoToleranceOnAttributes = [];
-        $unretrievableAttributes = [];
+        $attributesArray = [];
```
Comment (nunomaduro) [inline src/Settings/LocalFactory.php]: `array_fill_keys` || `array_fill`.

File: src/Settings/LocalFactory.php [hunk 4]
```diff
@@ -128,117 +70,26 @@ public function __construct(RemoteSettingsRepository $remoteRepository)
...
-                $customRanking[] = "desc({$key})";
-            }
-
-            if ($this->isDisableTypoToleranceOnAttributes($key, $value)) {
-                $disableTypoToleranceOnAttributes[] = $key;
-            }
-
-            if ($this->isUnretrievableAttributes($key, $value)) {
-                $unretrievableAttributes[] = $key;
+            foreach (self::$settings as $setting => $class) {
```
Comment (nunomaduro) [inline src/Settings/LocalFactory.php]: $settingClass

File: src/Contracts/SettingContract.php (hunk 1 above)
Comment (nunomaduro) [inline src/Contracts/SettingContract.php]: `arrayValue`.

File: src/Settings/LocalFactory.php [hunk 5]
```diff
@@ -36,73 +39,12 @@ final class LocalFactory
...
-     */
-    private static $disableTypoToleranceOnAttributesKeys = [
-        'slug',
-        '*_slug',
-        'slug_*',
-        '*code*',
-        '*sku*',
-        '*reference*',
+    private static $settings = [
+        'searchableAttributes' => UnsearcheableAttribute::class,
```
Comment (nunomaduro) [inline src/Settings/LocalFactory.php]: Same name

File: src/Settings/SettingAttribute/CustomRankingAttribute.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        'number_*',
+        '*_number',
+    ];
+
+    /**
+     * Checks if the given key/value is a 'CustomRanking'.
+     *
+     * @param  string $key
+     * @param  mixed $value
+     * @param array $customRanking
```
Comment (nunomaduro) [inline src/Settings/SettingAttribute/CustomRankingAttribute.php]: missing space between param and array.

File: src/Settings/SettingAttribute/CustomRankingAttribute.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+     *
+     * @return array
+     */
+    public static function exist(string $key, $value, array $customRanking): array
+    {
+        if (Str::is(self::$customRankingKeys, $key)) {
+            $customRanking[] = "desc({$key})";
+
```
Comment (nunomaduro) [inline src/Settings/SettingAttribute/CustomRankingAttribute.php]: Only one return

File: src/Settings/SettingAttribute/UnretrievableAttribute.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    ];
+
+    /**
+     * Checks if the given key/value is a 'UnretrieableAttribute'.
+     *
+     * @param  string $key
+     * @param  mixed $value
+     * @param array $unretrievableAttributes
+     *
+     * @return array
```
Comment (nunomaduro) [inline src/Settings/SettingAttribute/UnretrievableAttribute.php]: Go deeper on static analysis. `[]string`.

File: src/Settings/SettingAttribute/UnretrievableAttribute.php [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        '*token*',
+        '*secret*',
+        '*hash*',
+    ];
+
+    /**
+     * Checks if the given key/value is a 'UnretrieableAttribute'.
+     *
+     * @param  string $key
+     * @param  mixed $value
```
Comment (nunomaduro) [inline src/Settings/SettingAttribute/UnretrievableAttribute.php]: Specific on the value.

File: src/Contracts/SettingContract.php (hunk 1 above)
Comment (TomKlotzPro) [reply src/Contracts/SettingContract.php]: not sure i've did it correctly need review @nunomaduro

File: src/Contracts/SettingContract.php [hunk 10]
```diff
@@ -13,7 +13,22 @@
 
 namespace Algolia\ScoutExtended\Contracts;
 
+/**
+ * This interface is use to check if key exist.
```
Comment (nunomaduro) [inline src/Contracts/SettingContract.php]: remove this.

File: src/Settings/LocalFactory.php [hunk 11]
```diff
@@ -70,17 +70,17 @@ public function __construct(RemoteSettingsRepository $remoteRepository)
     public function create(SearchIndex $index, string $model): Settings
     {
         $attributes = $this->getAttributes($model);
-        $attributesArray = [];
+        $arrayFillKeys = [];
```
Comment (nunomaduro) [inline src/Settings/LocalFactory.php]: The comment was for *using* the array_fill_keys function. Not rename the variable.

Comment (TomKlotzPro) [reply src/Settings/LocalFactory.php]: sorry, my bad. it's done.

**Comment by TomKlotzPro** [comment]
Good @nunomaduro

**Comment by nunomaduro** [comment]
Merged into: https://github.com/algolia/scout-extended/pull/170

**Pull request closed without merging**